docs/gitbook/getting_started/tutorial.md

```diff
@@ -0,0 +1,493 @@
+| category#electronics | 0.7443372011184692 |
+| category#entertainment | 5.039264578954317e-07 |
+| category#food | 1.415687620465178e-07 |
+| category#sports | -0.2697771489620209 |
+| gender#male | 0.7336684465408325 |
+| gender#female | 0.47442761063575745 |
+| gender#other | 5.039264578954317e-07 |
+| price | -110.62307739257812 |
+
+Notice that weight is learned for each possible value in a categorical feature, and for every single quantitative feature.
+
+Of course, you can optimize hyper-parameters to build more accurate prediction model. Check the output of the following query to see all available options, including learning rate, number of iterations and regularization parameters, and their default values:
+
+```sql
+select train_classifier(array(), 0, '-help')
+```
+
+### Step 3. Prediction
+
+Now, the table `classifier` has liner coefficients for given features, and we can predict unforeseen samples by computing a weighted sum of their features.
+
+How about the probability of purchase by a `male` customer who sees a `food` product priced at `120` on `Friday`? Which product is more likely to be purchased by the customer on `Friday`?
+
+To differentiate potential purchases, create a `unforeseen_samples` table with these unknown combinations of features:
+
+```sql
+create table if not exists unforeseen_samples
+ (id bigint, features array<string>)
+;
+```
+
+```sql
+insert overwrite table unforeseen_samples
+select 1 as id, array("gender#male", "category#food", "day of week#Friday", "price:120") as features
+union all
+select 2 as id, array("gender#male", "category#sports", "day of week#Friday", "price:1000") as features
+union all
+select 3 as id, array("gender#male", "category#electronics", "day of week#Friday", "price:540") as features
+;
+```
+
+Prediction for the feature vectors can be made by join operation between `unforeseen_samples` and `classifier` on each feature as:
+
+```sql
+with features_exploded as (
+  select
+    id,
+    -- split feature string into its name and value
+    -- to join with a model table
+    extract_feature(fv) as feature,
+    extract_weight(fv) as value
+  from unforeseen_samples t1 LATERAL VIEW explode(features) t2 as fv
+)
+select
+  t1.id,
+  sigmoid( sum(p1.weight * t1.value) ) as probability
+from
+  features_exploded t1
+  LEFT OUTER JOIN classifier p1 ON (t1.feature = p1.feature)
+group by
+  t1.id
+;
+```
+
+Output for single sample can be:
+
+```
+id, probability
+ 1, 1.0261879540562902e-10
+```
+
+### Evaluation
+
+If you have test samples for evaluation, use Hivemall's [evaluation UDFs](http://hivemall.incubator.apache.org/userguide/eval/binary_classification_measures.html) to measure the accuracy of prediction.
+
+For instance, prediction accuracy over the `training` samples can be measured as:
+
+```sql
+with features_exploded as (
+  select
+    id,
+    extract_feature(fv) as feature,
+    extract_weight(fv) as value
+  from training t1 LATERAL VIEW explode(features) t2 as fv
```

Comment:
Now we didn't create `training` table above, so how about creating an alternative `testing` table and use it for evaluation?

```sql
create table testing as
select 1 as id, array("gender#female", "category#food", "day of week#Friday", "price:500") as features, 1 as label
union all
select 2 as id, array("gender#male", "category#food", "day of week#Saturday", "price:250") as features, 0 as label
union all
select 3 as id, array("gender#other", "category#electronics", "day of week#Thursday", "price:330") as features, 1 as label
;
```

```sql
with features_exploded as (
  select
    id,
    extract_feature(fv) as feature,
    extract_weight(fv) as value
  from testing t1 LATERAL VIEW explode(features) t2 as fv
),
predictions as (
  select
    t1.id,
    sigmoid( sum(p1.weight * t1.value) ) as probability
  from
    features_exploded t1
    LEFT OUTER JOIN classifier p1 ON (t1.feature = p1.feature)
  group by
    t1.id
)
select
  auc(probability, label) as auc,
  logloss(probability, label) as logloss
from (
  select t1.probability, t2.label
  from predictions t1
  join testing t2 on (t1.id = t2.id)
  ORDER BY probability DESC
) t
;
```

Reply:
Will use CTAS.